### PR TITLE
Port delegation tests to rust

### PR DESCRIPTION
**Pull request by frederikrothenberger**
Ports tests related to prepare_delegation and get_delegation to rust.
The tests were not ported 1:1 because some haskell tests did too many things, especially related to upgrades.
All testcases of the haskell tests should still be covered.

### CODE REVIEW

File: src/canister_tests/src/tests.rs [hunk 1]
```diff
@@ -66,6 +66,462 @@ fn registration_with_mismatched_sender_fails() {
...
+                    .add(Duration::from_secs(30 * 60)) // default expiration: 30 minutes
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_nanos() as u64
+            );
+            prepare_delegation_results.push((canister_sig_key, expiration))
+        }
+
+        for (idx, (session_key, frontend_hostname, _)) in delegation_params.iter().enumerate() {
+            let (canister_sig_key, expiration) = prepare_delegation_results.get(idx).unwrap();
```
Comment (nmattia) [inline src/canister_tests/src/tests.rs:278]: question: how do we make sure the delegation is ready here? when does the signing/certification happen?

File: src/canister_tests/src/tests.rs [hunk 2]
```diff
@@ -66,6 +66,462 @@ fn registration_with_mismatched_sender_fails() {
...
+
+            assert_eq!(
+                expiration,
+                env.time()
+                    .add(Duration::from_secs(30 * 60)) // default expiration: 30 minutes
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_nanos() as u64
+            );
+            prepare_delegation_results.push((canister_sig_key, expiration))
```
Comment (nmattia) [inline src/canister_tests/src/tests.rs:274]: why create `prepare_delegation_results` and not simply check the delegation here?

File: src/canister_tests/src/tests.rs [hunk 3]
```diff
@@ -66,6 +66,462 @@ fn registration_with_mismatched_sender_fails() {
...
+            };
+
+            framework::verify_delegation(&env, canister_sig_key.clone(), &signed_delegation);
+            assert_eq!(signed_delegation.delegation.pubkey, (*session_key).clone());
+            assert_eq!(signed_delegation.delegation.expiration, *expiration);
+        }
+        Ok(())
+    }
+
+    /// Verifies that an anchor that was registered using II_WASM_PREVIOUS gets valid delegations after upgrading to the current version.
```
Comment (nmattia) [inline src/canister_tests/src/tests.rs:299]: very nice

File: src/canister_tests/src/tests.rs (hunk 1 above)
Comment (frederikrothenberger) [reply src/canister_tests/src/tests.rs:278]: The `state_machine` executes ingress synchronously, i.e. the `prepare_delegation` update call will only return once it is complete. The whole `call` and then `read_state` polling has been cut away in this test infrastructure.

Comment (nmattia) [reply src/canister_tests/src/tests.rs:278]: gotcha, a short comment then would help

File: src/canister_tests/src/tests.rs (hunk 2 above)
Comment (nmattia) [reply src/canister_tests/src/tests.rs:274]: ah I see! then you can probably get rid of the intermediate structures + get rid of a lot of room for mishandling indices by doing something like this:

``` rust
        let foo = delegation_params.clone().into_iter().map(|(session_key, frontend_hostname, time)| {
            env.set_time(time);
            let (canister_sig_key, expiration) = api::prepare_delegation(
                &env,
                canister_id,
                principal_1(),
                user_number,
                frontend_hostname.to_string(),
                session_key.clone(),
                None,
            ).unwrap();

            assert_eq!(
                expiration,
                env.time()
                    .add(Duration::from_secs(30 * 60)) // default expiration: 30 minutes
                    .duration_since(UNIX_EPOCH)
                    .unwrap()
                    .as_nanos() as u64
            );
            (session_key, frontend_hostname, canister_sig_key, expiration)
        });

        for (session_key, frontend_hostname, canister_sig_key, expiration) in foo {
            let signed_delegation = match api::get_delegation(
                &env,
                canister_id,
                principal_1(),
                user_number,
                frontend_hostname.to_string(),
                (*session_key).clone(),
                expiration,
            )? {
                GetDelegationResponse::SignedDelegation(delegation) => delegation,
                GetDelegationResponse::NoSuchDelegation => panic!("failed to get delegation"),
            };

            framework::verify_delegation(&env, canister_sig_key.clone(), &signed_delegation);
            assert_eq!(signed_delegation.delegation.pubkey, (*session_key).clone());
            assert_eq!(signed_delegation.delegation.expiration, expiration);
        };
```

**Review by nmattia — APPROVED** [review]

**Pull request merged**